src/components/Firestore/Field/DocumentStore.tsx

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from 'react';
+import get from 'lodash.get';
+import produce from 'immer';
+import { Action, createReducer } from 'typesafe-actions';
+
+import { isMap, isArray, getLeafPath, getParentPath } from './utils';
+import { FirestoreAny } from './models';
+import * as actions from './actions';
+
+const reducer = createReducer<FirestoreAny, Action>({})
+  .handleAction(actions.reset, (state, { payload }) => {
+    return payload;
+  })
+  .handleAction(
+    actions.addField,
+    produce((draft: FirestoreAny, { payload }) => {
+      const parent = get(draft, getParentPath(payload.path)) || draft;
+      if (isMap(parent)) {
+        parent[getLeafPath(payload.path)] = payload.value;
+      } else if (isArray(parent)) {
+        parent.push(payload.value);
+      }
```

Comment:
What happens if parent is string, null, etc.?

Reply:
It should never happen... we can assume and the types are defined such that only FirestoreArray and FirestoreMap can have child elements in the UI